06-security.md

```diff
@@ -1 +1,451 @@
+An XSRF token is assigned to the browser on the first GET request. XSRF cookies MUST have the flag `secure` set. The API must check PUT, POST, PATCH and DELETE requests for the presence of the correct XSRF token.
+
+### Authentication
+
+Usernames and passwords are still the most commonly used way to authenticate a user especially if there is no supporting infrastructure already in place. However implementing proper authentication using usernames and passwords is not a trivial task. Owasp has a [whole page](https://www.owasp.org/index.php/Authentication_Cheat_Sheet) on it. This convention describes just the minimal requirements for authenticating using a username and password. Other (important) subjects such as 'Forgotten Password' functionality or account activation is out of scope. 
+
+#### Passwords must be salted and hashed using BCrypt
+
+BCrypt is a computational heavy hashing algorithm with built in salt. If the user database is stolen it will take a lot more effort to brute force the passwords than by using conventional algorithms (MD5, SHA-1, SHA-256). Note that there is a better hashing algorithm called Argon2 but this is not (yet) supported by Spring-Security.  
+
+#### A failed login attempt must always return the same response.
+
+When authentication fails the username may not exist, the password may be incorrect or the account may have been locked. If an error message is returned it should always be a generic error message ('Authentication Failed') that does not state the reason for failing. This prevents attackers from learning if actually an account exists with that username.
+
+#### Limit the number of Login attempts per unit of time.
+
+If the login form is accessible from the Internet, the number of (failed) login attempts must be limited per unit of time to prevent brute forcing. The counting must be done in the user account, not in the session (as a new session may be used for each attempt). 
+
+#### Consider using password complexity requirements and blacklists
+
+While having a these will not prevent [pattern based brute forcing techniques](https://www.youtube.com/watch?v=zUM7i8fsf0g) it will prevent users from picking too simple or common passwords.
+Good blacklists can be found [here](https://github.com/danielmiessler/SecLists/tree/master/Passwords). 
+
+### Single Origin Policy issues
+
+If a browser application that resides on a different domain access needs to access the API, the Single Origin Policy (SOP) will prevent the browser from reading the responses or block the request entirely. There are various ways of working around the SOP:
+
+- add Cross Origin Resource Sharing headers to the API.
+- use an API Proxy on the origin host.
+
+Please note that if both the single page web application and API are under your control its often more convenient to host both on the same physical server and map their end-points onto a different URL in the same domain. 
+
+#### Implement Cross Origin Resource Sharing if needed
+
```

Comment:
I would like to see a paragraph here that states CORS is often not required for regular FE-BE applications. In most cases, using a /api convention which is routed by Apache to the Tomcat server suffices. Too often it happened that people work from the presumption of CORS (thank you, StackOverflow), whereas the better alternative was not known.